Natapp.go

```diff
@@ -42,97 +43,129 @@ func init() {
+func processPacket(tupleV TupleV) TupleV {
+	var iPAndPort IPAndPort
+	tupleVKey := createVTupleString(tupleV)
+	if iPAndPort,isIn := readFromUsedPorts(tupleVKey, session); !isIn{
+	//if (IPAndPort{} == iPAndPort) {
+		iPAndPort = selectIPAndPort()
 		updateIpsAndPorts(iPAndPort)
-		updateUsedPorts(tuple5Key, iPAndPort);
-		reverseTuple5 := extractReverseTuple(tuple5, iPAndPort)
-		sourceIpAndPort := IPAndPort{destinationIp: tuple5.sourceIp, destinationPort: tuple5.sourcePort}
-		updateUsedPorts(create5TupleKey(reverseTuple5),sourceIpAndPort)
+		updateUsedPorts(tupleVKey, iPAndPort)
+
+		reverseTupleV := extractReverseTuple(tupleV, iPAndPort)
+		sourceIpAndPort := IPAndPort{destinationIp: tupleV.sourceIp, destinationPort: tupleV.sourcePort}
+		updateUsedPorts(createVTupleString(reverseTupleV),sourceIpAndPort)
 	}
-	tuple5.destinationIp = iPAndPort.destinationIp
-	tuple5.destinationPort = iPAndPort.destinationPort
-	return tuple5
+	tupleV.destinationIp = iPAndPort.destinationIp
+	tupleV.destinationPort = iPAndPort.destinationPort
+	return tupleV
 }
 
-func create5TupleKey(tuple5 Tuple5) string {
-	return create5TupleString(tuple5,"")
-}
-func create5TupleString(tuple5 Tuple5, seperator string) string{
-	string := strings.Join([]string{tuple5.sourceIp, strconv.Itoa(tuple5.sourcePort), tuple5.destinationIp, strconv.Itoa(tuple5.destinationPort), strconv.Itoa(tuple5.protocol)}, seperator)
-	return string
+func createVTupleString(tupleV TupleV) string{
+	res := strings.Join([]string{tupleV.sourceIp, strconv.Itoa(tupleV.sourcePort), tupleV.destinationIp, strconv.Itoa(tupleV.destinationPort), strconv.Itoa(tupleV.protocol)}, SEPERATOR)
+	return res
 }
 
-func readFromUsedPorts(tuple5Key string, session *gocql.Session) IPAndPort {
-	var ipAndPort IPAndPort
+func readFromUsedPorts(tuple5Key string, session *gocql.Session) (IPAndPort,bool) {
+	var flag bool
+	ipAndPort := IPAndPort{destinationIp: "", destinationPort: 0, used : true}
 	arg := fmt.Sprintf("SELECT ip,port FROM usedports WHERE key=?")
 	if err := session.Query(arg, tuple5Key).Scan(&ipAndPort.destinationIp, &ipAndPort.destinationPort); err != nil {
-		log.Fatal(err)
+		fmt.Println(err)
+		flag = false
+	} else {
+		flag = ipAndPort.destinationIp == ""
 	}
-	return ipAndPort
+	return ipAndPort, flag
 }
 
 func selectIPAndPort() IPAndPort {
-	ipsAndPorts := getIPsAndPorts();
-	if (len(ipsAndPorts) == 0) {
-		panic("There is no remaning ip ports")
-	}
-	index := rand.Int() % len(ipsAndPorts)
-
-	return ipsAndPorts[index]
-}
-
-func getIPsAndPorts() []IPAndPort {
-	var ipsAndPorts []IPAndPort
+	var ipAndPort IPAndPort
 	var ipAndPortStr string
-	var ipUsed bool
-	var i = 0
-	iter := session.Query("SELECT * from ports").Iter()
+	var flag= false
+
+	iter := session.Query("SELECT * FROM ports").Iter()
 	for {
-		i++;
 		row := map[string]interface{}{
-			"key":   &ipAndPortStr,
-			"used": &ipUsed,
+			"key":  &ipAndPortStr,
+			"used": &ipAndPort.used,
 		}
 		if !iter.MapScan(row) {
```

Comment:
Here we can just have the panic message:"There is no remaining ip and ports". No need to add a flag